File: person_task_manager/server/src/ui/routers/group-task.router.ts
import { NextFunction, Request, Response, Router } from "express";
import { RequestValidator } from "../../core/common/error-handler";
import { GroupTaskRequestDto } from "../../core/domain/dtos/group-task.dto";
import { updateNameRequestDto } from "../../core/domain/dtos/request_dtos/update-name-request.dto";
import { groupTaskController } from "../controllers/group-task.controller";
import { returnResult } from "../../kernel/util/return-result";
import { ARCHIVE_GROUP_TASK_FAILED, CREATE_GROUP_TASK_FAILED, DELETE_GROUP_TASK_FAILED, ENABLE_GROUP_TASK_FAILED, GROUP_TASK_NOT_FOUND, TASK_NO_RECORDS, UPDATE_GROUP_TASK_FAILED } from "../../core/domain/constants/error.constant";
import { taskController } from "../controllers/task.controller";

export const groupTaskRouter = Router();

const groupTaskControllerImpl = groupTaskController;
const taskControllerImpl = taskController;

// get one group task
groupTaskRouter.get("/:id", async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
        const groupTaskResult = await groupTaskControllerImpl.getGrouptaskById(req, next);
        returnResult(groupTaskResult, GROUP_TASK_NOT_FOUND, res, next);
    }
    catch (err) {
        next(err);
    }
});

// create group task
groupTaskRouter.post("/create",
    RequestValidator.validateV2(GroupTaskRequestDto),
    async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
        const groupTaskResult = await groupTaskControllerImpl.createGroupTask(req, next);
        returnResult(groupTaskResult, CREATE_GROUP_TASK_FAILED, res, next);
    }
    catch (err) {
        next(err);
    }
});

// update group task
groupTaskRouter.put("/:id", 
    RequestValidator.validateV2(GroupTaskRequestDto),
    async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
        const groupTaskResult = await groupTaskControllerImpl.updateGroupTask(req, next);
        returnResult(groupTaskResult, UPDATE_GROUP_TASK_FAILED, res, next);
    }
    catch (err) {
        next(err);
    }
});

// delete group task
groupTaskRouter.delete("/:id", async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
        const groupTaskResult = await groupTaskControllerImpl.deleteGroupTask(req, next);
        returnResult(groupTaskResult, DELETE_GROUP_TASK_FAILED, res, next);
    }
    catch (err) {
        next(err);
    }
});

// get all tasks of a group task
groupTaskRouter.get("/:id/tasks", async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
        const groupTaskResult = await taskControllerImpl.getTasksByGroupTaskId(req, next);
        returnResult(groupTaskResult, GROUP_TASK_NOT_FOUND, res, next);
    }
    catch (err) {
        next(err);
    }
});

// update group task name
groupTaskRouter.put("/:id/update-name", 
    RequestValidator.validateV2(updateNameRequestDto),
    async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
        const groupTaskResult = await groupTaskControllerImpl.updateGroupTaskName(req, next);
        returnResult(groupTaskResult, UPDATE_GROUP_TASK_FAILED, res, next);
    }
    catch (err) {
        next(err);
    }
});

// calculate total tasks and total tasks completed
groupTaskRouter.get("/:id/tasks-complete", async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
        const groupTaskResult = await groupTaskControllerImpl.calculateCompletedTasks(req, next);
        returnResult(groupTaskResult, TASK_NO_RECORDS, res, next);
    }
    catch (err) {
        next(err);
    }
});

// update ordinal number
groupTaskRouter.put("/:id/update-ordinal", async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
        const groupTaskResult = await groupTaskControllerImpl.updateOrdinalNumber(req, next);
        returnResult(groupTaskResult, UPDATE_GROUP_TASK_FAILED, res, next);
    }
    catch (err) {
        next(err);
    }
});

// archive group task
groupTaskRouter.put("/:id/archive", async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
        const groupTaskResult = await groupTaskControllerImpl.archiveGroupTask(req, next);
        returnResult(groupTaskResult, ARCHIVE_GROUP_TASK_FAILED, res, next);
    }
    catch (err) {
        next(err);
    }
});

// enable group task
groupTaskRouter.put("/:id/enable", async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
        const groupTaskResult = await groupTaskControllerImpl.enableGroupTask(req, next);
        returnResult(groupTaskResult, ENABLE_GROUP_TASK_FAILED, res, next);
    }
    catch (err) {
        next(err);
    }
});

File: middleware_loader/ui/routers/task_manager/group_task_router.go
package routers

import (
	"middleware_loader/core/domain/enums"
	"middleware_loader/core/middleware"
	services "middleware_loader/core/services/task_manager"
	database_mongo "middleware_loader/kernel/database/mongo"
	"middleware_loader/ui/controller_services"
	"net/http"

	"github.com/go-chi/chi"
)

type GroupTaskRouter struct {
	GroupTaskService *services.GroupTaskService
}

func NewGroupTaskRouter(groupTaskService *services.GroupTaskService, db database_mongo.Database, r *chi.Mux) *GroupTaskRouter {
	r.Route("/group-task", func(r chi.Router) {
		r.Use(middleware.CheckMicroserviceStatus(db, enums.TASK_MANAGER))
		r.Get("/{id}", func(w http.ResponseWriter, r *http.Request) {
			controller_services.GetGroupTaskById(w, r, groupTaskService)
		})
		r.Post("/create", func(w http.ResponseWriter, r *http.Request) {
			controller_services.CreateGroupTask(w, r, groupTaskService)
		})
		r.Put("/{id}", func(w http.ResponseWriter, r *http.Request) {
			controller_services.UpdateGroupTask(w, r, groupTaskService)
		})
		r.Delete("/{id}", func(w http.ResponseWriter, r *http.Request) {
			controller_services.DeleteGroupTask(w, r, groupTaskService)
		})
		r.Get("/{id}/tasks", func(w http.ResponseWriter, r *http.Request) {
			controller_services.GetTasksByGroupTask(w, r, groupTaskService)
		})
		r.Put("/{id}/update-name", func(w http.ResponseWriter, r *http.Request) {
			controller_services.UpdateGroupTaskName(w, r, groupTaskService)
		})
		r.Get("/{id}/tasks-complete", func(w http.ResponseWriter, r *http.Request) {
			controller_services.CalculateCompletedTasks(w, r, groupTaskService)
		})
		r.Put("/{id}/update-ordinal", func(w http.ResponseWriter, r *http.Request) {
			controller_services.UpdateGroupTaskOrdinal(w, r, groupTaskService)
		})
		r.Put("/{id}/archive", func(w http.ResponseWriter, r *http.Request) {
			controller_services.ArchiveGroupTask(w, r, groupTaskService)
		})
		r.Put("/{id}/enable", func(w http.ResponseWriter, r *http.Request) {
			controller_services.EnableGroupTask(w, r, groupTaskService)
		})
	})
	return &GroupTaskRouter{
		GroupTaskService: groupTaskService,
	}
}
